### DISCUSSION TITLE
About best practices for implementing loss functions

### DISCUSSION THREAD

**Discussion by minasora** [open-mmlab/mmengine#1139] [Q&A] [answered]
Thank you for your excellent work. I want to ask for the best practices for the loss function. I have seen in the Get Start, the loss function is in the model class:
```
        if mode == 'loss':
            return {'loss': F.cross_entropy(x, labels)}
```

How should I separate the loss function and set it in the config just like mmdet?
```
   loss_cls=dict(
            type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0),
```
 What is the best practice for the position of the loss function in mmengine? I searched the documentation but couldn't find an answer. Could you please enlighten me?

**Comment by narthchin** [comment, marked as answer]
Take an example from `mmdet`:
https://github.com/open-mmlab/mmdetection/blob/main/mmdet/models/dense_heads/anchor_head.py
Similarly, you can add `loss_cls` and more in your model `__init__` func and use `MODELS.build` to build the loss func from your config file, then in your `forward` function or somewhere else:
```python
losses_cls = self.loss_cls(pred, gt)
return dict(loss_cls=losses_cls)
```

**Reply by minasora** [reply]
Thank for your reply!
